demo/bulk_insert/bulk_insert.py

```diff
@@ -2,74 +2,195 @@
 import csv
 import os
 import click
-
-def ProcessNodes(LABELS, NODES, nodes_csv_files):
-	# Introduce nodes
-	LABELS.append("NODES")
-	NODE_COUNT_IDX = len(LABELS)
-	NODES_TO_CREATE = 0
-	LABELS.append(NODES_TO_CREATE)
-
-	# Number of labels
-	LABELS.append(len(nodes_csv_files))
-
+# TODO for debugging only; remove
+import ipdb
+
+class Argument:
+    def __init__(self, argtype):
+	self.argtype = argtype
+	self.pending_inserts = 0
+	self.type_count = 0
+	self.descriptors = []
+
+    def reset_tokens(self):
+	self.pending_inserts = 0
+        for descriptor in self.descriptors:
+	    descriptor.pending_inserts = 0
+
+    def remove_descriptors(self, delete_count):
+	del self.descriptors[:delete_count]
+	self.type_count -= delete_count
+
+    def unroll(self):
+	ret = [self.argtype, self.pending_inserts, self.type_count]
+        for descriptor in self.descriptors:
+	    # Don't include a descriptor unless we are inserting its entities.
+	    # This can occur if the addition of a descriptor and its first entity caused
+	    # the token count to exceed the max allowed
+	    if descriptor.pending_inserts == 0:
+		ret[2] -= 1
+	    else:
+		ret += descriptor.unroll()
+	return ret
+
+    def token_count(self):
+	# ["NODES"/"RELATIONS", number of entities to insert, number of different entity types, list of type descriptors]
+	return 3 + sum(descriptor.token_count() for descriptor in self.descriptors)
+
+# Structure of a node insertion:
+# ["NODES", node count, label count, label_descriptors[0..n]]
+class Descriptor:
+    def __init__(self, name):
+	self.name = name
+	self.pending_inserts = 0
+
+# Structure of a label descriptor:
+# [label name, node count, attribute count, attributes[0..n]]
+class LabelDescriptor(Descriptor):
+    def __init__(self, name):
+	self.name = name
+	self.pending_inserts = 0
+	self.attribute_count = 0
+	self.attributes = []
+
+    def token_count(self):
+	# Labels have a token for name, attribute count, pending insertion count, plus N tokens for the individual attributes.
+	return 3 + self.attribute_count
+
+    def unroll(self):
+	return [self.name, self.pending_inserts, self.attribute_count] + self.attributes
+
+# TODO This seems kind of unnecessary, and should maybe be refactored
+class RelationDescriptor(Descriptor):
+    def token_count(self):
+	# Relations have 2 tokens: name and pending insertion count.
+	return 2
+
+    def unroll(self):
+	return [self.name, self.pending_inserts]
+
+MAX_TOKENS = 1024 * 1024
+graphname = None
+redis_client = None
+
+def QueryRedis(metadata, entities):
+    cmd = metadata.unroll() + entities
+    # TODO - asynchronous execution of these queries would be really nice; we don't need to
+    # wait for a response. These two approaches take about the same time, though -
+    # find a solution
```

Comment:
Let me know if you have any thoughts here!

Reply:
pipeline didn't improved creation time?

Reply:
As implemented here, the timings were almost identical when inserting a 62-megabyte Label csv file (~300k nodes, ~5 million tokens, 5 calls to Redis).

Pipelined:
```
real    0m4.968s
user    0m3.780s
sys     0m0.109s
```
Using `execute_command`:
```
real    0m4.989s
user    0m3.774s
sys     0m0.233s
```

I believe that `pipe.execute()` is still a blocking call - Redis doesn't return between queries within a single `execute` call, but Python still waits until the full set of queries finishes. In our case, I would like to send the Redis query and then continue processing the input file - as long as the queries are processed in order, we don't need their responses within Python (except for node and edge creation counts, which doesn't seem hugely important to me).

With the example here, the input processing (without sending any data to Redis) takes ~0.38 seconds, so a very crude estimation suggests that the maximum speed benefit of processing instead of waiting for replies would be a 7.7% speedup.

Reply:
Calling `pipe.execute()` once:
```
real    0m5.014s
user    0m3.773s
sys     0m0.112s
```
I feel like this approach makes less sense for us, as it requires the complete input to be stored in Python memory and Redis memory simultaneously.

Reply:
Agree.

Reply:
One thought I have is that we can place each invocation of the `QueryRedis` function in a child thread, but at the moment I don't think the potential speed gains make that worthwhile.

If we later elect to make the Python processing operations more demanding - such as by representing relations by the connected entities rather than Node IDs - it might make more sense to run insert and processing operations in parallel.